experimental/sdk/webpubsub-socketio-extension/src/SIO/components/web-pubsub-adapter.ts

```diff
@@ -171,6 +252,57 @@ export class WebPubSubAdapterInternal extends NativeInMemoryAdapter {
    * @param close - whether to close the underlying connection
    */
   public disconnectSockets(opts: BroadcastOptions, close: boolean): void {
-    throw new Error("Not implemented");
+    throw NotSupportedError;
+  }
+  
+  /**
+   * Generates OData filter string for Web PubSub service from a set of rooms and a set of exceptions
+   * @param rooms - a set of Rooms to include
+   * @param except - a set of Rooms to exclude
+   * @returns OData - filter string
+   */
+  private _buildODataFilter(rooms: Set<string>, excepts: Set<string> | undefined): string {
+    debug("_buildODataFilter");
+    let allowFilter = "";
+    let room_idx = 0,
+      except_idx = 0;
+
+    if (rooms.size === 0) rooms = new Set([""]);
+    for (const room of rooms) {
+      const groupName = this._getGroupName(this.nsp.name, room);
+      allowFilter += `'${groupName}' in groups` + (room_idx === rooms.size - 1 ? "" : " or ");
+      room_idx++;
+    }
+
+    let denyFilter = "";
+    if (excepts)
+      {for (const except of excepts) {
+        const exceptGroupName = this._getGroupName(this.nsp.name, except);
+        denyFilter +=
+          `not ('${exceptGroupName}' in groups)` + (except_idx === excepts.size - 1 ? "" : " and ");
+        except_idx++;
+      }}
+
+    let result = "";
+    if (allowFilter.length > 0)
+      {result = allowFilter + (denyFilter.length > 0 ? " and " + denyFilter : "");}
+    else result = denyFilter.length > 0 ? `${denyFilter}` : "";
+    debug(`_buildODataFilter result = ${result}`);
+    return result;
+  }
+
+  
+  private _getEioSid(sioSid: string): string {
+    return (this.nsp.sockets.get(sioSid).conn as any).id;
+  }
+
+  /**
+   * `namespace` and `room` are concpets from Socket.IO.
+   * `group` is a concept from Azure Web PubSub.
+   */
+  private _getGroupName(namespace: string, room?: string): string {
```

Comment:
is it empty or undefined standing for *all groups*?

Reply:
it stands for a single and special group, which stands for the whole namespace.